{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## UMLS-JKG Extract Scripts\n",
    "#### Assumes accessible Oracle database of UMLS Metathesauras and Semantic Network\n",
    "#### Generates a JKG file for ingest into neo4j - entire script runs a bit over an hour and creates 4 to 5 GB file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import json as json\n",
    "import cx_Oracle\n",
    "import sqlalchemy\n",
    "pd.set_option('display.max_rows', 10)\n",
    "pd.set_option('display.max_colwidth', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Establish a connection - hide password in file home directory\n",
    "##### conn_string.txt file contains one line like: oracle+cx_oracle://user:pass@server-address:port/database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = open('conn_string.txt', 'r'); conn_string = _.read().replace('\\n',''); _.close()\n",
    "engine = sqlalchemy.create_engine(conn_string, arraysize=100000, max_identifier_length=128, encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Set the UMLSversion (in Pitt/Neptune implementation this is the Oracle Schema/User)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "UMLSversion = 'UMLS2025AB'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Concept-Concept rels from MRREL, MRSAB and inverse file - done first due to being most likely to fail and used twice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_id</th>\n",
       "      <th>end_id</th>\n",
       "      <th>rel_label</th>\n",
       "      <th>sab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C0628348</td>\n",
       "      <td>C0000097</td>\n",
       "      <td>mapped_to</td>\n",
       "      <td>MSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0019665</td>\n",
       "      <td>C0000172</td>\n",
       "      <td>aq</td>\n",
       "      <td>MSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0045031</td>\n",
       "      <td>C0000173</td>\n",
       "      <td>mapped_to</td>\n",
       "      <td>MSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C2973356</td>\n",
       "      <td>C0000194</td>\n",
       "      <td>mapped_to</td>\n",
       "      <td>MSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0069813</td>\n",
       "      <td>C0000204</td>\n",
       "      <td>mapped_to</td>\n",
       "      <td>MSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12253742</th>\n",
       "      <td>C5220253</td>\n",
       "      <td>C6021589</td>\n",
       "      <td>has_loinc_number</td>\n",
       "      <td>LNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12253743</th>\n",
       "      <td>C6018018</td>\n",
       "      <td>C6021605</td>\n",
       "      <td>has_evaluation</td>\n",
       "      <td>LNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12253744</th>\n",
       "      <td>C6018252</td>\n",
       "      <td>C6021871</td>\n",
       "      <td>has_divisor</td>\n",
       "      <td>LNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12253745</th>\n",
       "      <td>C6021460</td>\n",
       "      <td>C6021878</td>\n",
       "      <td>has_system</td>\n",
       "      <td>LNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12253746</th>\n",
       "      <td>C1544498</td>\n",
       "      <td>C6022213</td>\n",
       "      <td>analyzes</td>\n",
       "      <td>LNC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12253747 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          start_id    end_id         rel_label  sab\n",
       "0         C0628348  C0000097         mapped_to  MSH\n",
       "1         C0019665  C0000172                aq  MSH\n",
       "2         C0045031  C0000173         mapped_to  MSH\n",
       "3         C2973356  C0000194         mapped_to  MSH\n",
       "4         C0069813  C0000204         mapped_to  MSH\n",
       "...            ...       ...               ...  ...\n",
       "12253742  C5220253  C6021589  has_loinc_number  LNC\n",
       "12253743  C6018018  C6021605    has_evaluation  LNC\n",
       "12253744  C6018252  C6021871       has_divisor  LNC\n",
       "12253745  C6021460  C6021878        has_system  LNC\n",
       "12253746  C1544498  C6022213          analyzes  LNC\n",
       "\n",
       "[12253747 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The following query was used to return all UMLS relationship types and counts to then manually determine inverse relationships\n",
    "# select lower(NVL(RELA, REL)) as rel_type, count(*) as count from UMLS2025AB.MRREL GROUP BY lower(NVL(RELA, REL)) ORDER BY count DESC\n",
    "# The script depends upon the manually managed file: UMLS_inverse_relationships.csv\n",
    "# This excludes inverse_relationships - query runs about a minute\n",
    "inverse_relationships = pd.read_csv('UMLS_inverse_relationships.csv')\n",
    "inverse_relationships.to_sql('inverse_rel', con=engine, if_exists='replace', index=False)\n",
    "query = \"WITH rels AS (SELECT DISTINCT CUI2, CUI1, LOWER(TO_CHAR(NVL(RELA, REL))) as relarel, SAB as relsab FROM {0}.MRREL WHERE SUPPRESS <> 'O' AND CUI1 <> CUI2), inverse AS (SELECT lower(TO_CHAR(rel_label)) as inv_rel FROM inverse_rel), SABlist AS (SELECT DISTINCT RSAB AS listsab from {0}.MRSAB WHERE LAT = 'ENG') SELECT CUI2, CUI1, relarel, relsab FROM rels INNER JOIN SABlist ON rels.relsab = SABlist.listsab LEFT JOIN inverse ON rels.relarel = inverse.inv_rel WHERE inverse.inv_rel IS NULL\".format(UMLSversion)\n",
    "df = pd.read_sql_query(query, engine)\n",
    "df.columns =['start_id','end_id','rel_label','sab']\n",
    "# replace dashes in relationship labels with underscore\n",
    "df['rel_label'] = df['rel_label'].str.replace('-','_')\n",
    "concept_rels = df\n",
    "concept_rels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create initial file and first line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('UMLS-JKG.json', 'w')\n",
    "f.write('{\"nodes\":[\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add UMLS and UMLS SABs as Sources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vsab</th>\n",
       "      <th>sab</th>\n",
       "      <th>name</th>\n",
       "      <th>srl</th>\n",
       "      <th>ttyl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AIR93</td>\n",
       "      <td>AIR</td>\n",
       "      <td>AI/RHEUM, 1993</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALT2009</td>\n",
       "      <td>ALT</td>\n",
       "      <td>Alternative Billing Concepts, 2009</td>\n",
       "      <td>3</td>\n",
       "      <td>AB,HT,PT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AOD2000</td>\n",
       "      <td>AOD</td>\n",
       "      <td>Alcohol and Other Drug Thesaurus, 2000</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AOT2003</td>\n",
       "      <td>AOT</td>\n",
       "      <td>Authorized Osteopathic Thesaurus, 2003</td>\n",
       "      <td>0</td>\n",
       "      <td>ET,PT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATC_2025_02_10_25_09_02</td>\n",
       "      <td>ATC</td>\n",
       "      <td>Anatomical Therapeutic Chemical Classification System, 2025_02_10</td>\n",
       "      <td>0</td>\n",
       "      <td>IN,PT,RXN_IN,RXN_PT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>USP_2025_01_30_25_09_02</td>\n",
       "      <td>USP</td>\n",
       "      <td>USP Compendial Nomenclature, 2025_01_30</td>\n",
       "      <td>0</td>\n",
       "      <td>CD,IN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>USPMG_2023</td>\n",
       "      <td>USPMG</td>\n",
       "      <td>USP Medicare Model Guidelines, 2023</td>\n",
       "      <td>0</td>\n",
       "      <td>HC,PT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>UWDA173</td>\n",
       "      <td>UWDA</td>\n",
       "      <td>University of Washington Digital Anatomist, 1.7.3</td>\n",
       "      <td>0</td>\n",
       "      <td>PT,SY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>VANDF_2025_07_31</td>\n",
       "      <td>VANDF</td>\n",
       "      <td>Veterans Health Administration National Drug File, 2025_07_31</td>\n",
       "      <td>0</td>\n",
       "      <td>AB,CD,IN,MTH_RXN_CD,PT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>WHO97</td>\n",
       "      <td>WHO</td>\n",
       "      <td>WHO Adverse Reaction Terminology, 1997</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>105 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        vsab    sab  \\\n",
       "0                      AIR93    AIR   \n",
       "1                    ALT2009    ALT   \n",
       "2                    AOD2000    AOD   \n",
       "3                    AOT2003    AOT   \n",
       "4    ATC_2025_02_10_25_09_02    ATC   \n",
       "..                       ...    ...   \n",
       "100  USP_2025_01_30_25_09_02    USP   \n",
       "101               USPMG_2023  USPMG   \n",
       "102                  UWDA173   UWDA   \n",
       "103         VANDF_2025_07_31  VANDF   \n",
       "104                    WHO97    WHO   \n",
       "\n",
       "                                                                  name  srl  \\\n",
       "0                                                       AI/RHEUM, 1993    0   \n",
       "1                                   Alternative Billing Concepts, 2009    3   \n",
       "2                               Alcohol and Other Drug Thesaurus, 2000    0   \n",
       "3                               Authorized Osteopathic Thesaurus, 2003    0   \n",
       "4    Anatomical Therapeutic Chemical Classification System, 2025_02_10    0   \n",
       "..                                                                 ...  ...   \n",
       "100                            USP Compendial Nomenclature, 2025_01_30    0   \n",
       "101                                USP Medicare Model Guidelines, 2023    0   \n",
       "102                  University of Washington Digital Anatomist, 1.7.3    0   \n",
       "103      Veterans Health Administration National Drug File, 2025_07_31    0   \n",
       "104                             WHO Adverse Reaction Terminology, 1997    2   \n",
       "\n",
       "                       ttyl  \n",
       "0                      None  \n",
       "1                  AB,HT,PT  \n",
       "2                      None  \n",
       "3                     ET,PT  \n",
       "4       IN,PT,RXN_IN,RXN_PT  \n",
       "..                      ...  \n",
       "100                   CD,IN  \n",
       "101                   HC,PT  \n",
       "102                   PT,SY  \n",
       "103  AB,CD,IN,MTH_RXN_CD,PT  \n",
       "104                    None  \n",
       "\n",
       "[105 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "f.write('{\"labels\":[\"Source\"],\"properties\":{\"id\":\"UMLS:UMLS\",\"name\":\"Unified Medical Language System\",\"description\":\"United States National Institutes of Health (NIH) National Library of Medicine (NLM) Unified Medical Language System (UMLS) Knowledge Sources.\",\"sab\":\"UMLS\" ,\"source\":\"http://www.nlm.nih.gov/research/umls/licensedcontent/umlsknowledgesources.html\"}}\\n')\n",
    "\n",
    "# Add NDC codes as source\n",
    "f.write(',{\"labels\":[\"Source\"] ,\"properties\":{\"id\":\"UMLS:NDC\" ,\"name\":\"National Drug Codes\",\"sab\":\"NDC\"}}\\n')\n",
    "    \n",
    "# Used info in MRSAB to add source restriction level (SRL) and term types (TTYL) to Source properties -  limit to current version (CURVER)\n",
    "query = \"SELECT VSAB, RSAB, SON, SRL, TTYL from {0}.MRSAB where {0}.MRSAB.LAT = 'ENG' AND CURVER = 'Y' ORDER BY RSAB\".format(UMLSversion)\n",
    "df = pd.read_sql_query(query, engine)\n",
    "df.columns =['vsab','sab','name','srl','ttyl']\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    f.write(',{\"labels\":[\"Source\"],\"properties\":{\"id\":\"UMLS:'+row[\"vsab\"]+'\",\"name\":'+json.dumps(row[\"name\"])+',\"sab\":\"'+row[\"sab\"]+'\",\"srl\":'+str(row[\"srl\"])+',\"ttyl\":'+json.dumps(row[\"ttyl\"])+'}}\\n')\n",
    "\n",
    "f.close()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Node_Labels from UMLS Semantic Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>def</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T012</td>\n",
       "      <td>Bird</td>\n",
       "      <td>A vertebrate having a constant body temperature and characterized by the presence of feathers.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>A normal or pathological part of the anatomy or structural organization of an organism.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T019</td>\n",
       "      <td>Congenital Abnormality</td>\n",
       "      <td>An abnormal structure, or one that is abnormal in size or location, present at birth or evolving...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T021</td>\n",
       "      <td>Fully Formed Anatomical Structure</td>\n",
       "      <td>An anatomical structure in a fully formed organism; in mammals, for example, a structure in the ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T048</td>\n",
       "      <td>Mental or Behavioral Dysfunction</td>\n",
       "      <td>A clinically significant dysfunction whose major manifestation is behavioral or psychological. T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>T120</td>\n",
       "      <td>Chemical Viewed Functionally</td>\n",
       "      <td>A chemical viewed from the perspective of its functional characteristics or pharmacological acti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>T129</td>\n",
       "      <td>Immunologic Factor</td>\n",
       "      <td>A biologically active substance whose activities affect or play a role in the functioning of the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>T169</td>\n",
       "      <td>Functional Concept</td>\n",
       "      <td>A concept which is of interest because it pertains to the carrying out of a process or activity.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>T170</td>\n",
       "      <td>Intellectual Product</td>\n",
       "      <td>A conceptual entity resulting from human endeavor. Concepts assigned to this type generally refe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>T192</td>\n",
       "      <td>Receptor</td>\n",
       "      <td>A specific structure or site on the cell surface or within its cytoplasm that recognizes and bin...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                               name  \\\n",
       "0    T012                               Bird   \n",
       "1    T017               Anatomical Structure   \n",
       "2    T019             Congenital Abnormality   \n",
       "3    T021  Fully Formed Anatomical Structure   \n",
       "4    T048   Mental or Behavioral Dysfunction   \n",
       "..    ...                                ...   \n",
       "122  T120       Chemical Viewed Functionally   \n",
       "123  T129                 Immunologic Factor   \n",
       "124  T169                 Functional Concept   \n",
       "125  T170               Intellectual Product   \n",
       "126  T192                           Receptor   \n",
       "\n",
       "                                                                                                     def  \n",
       "0         A vertebrate having a constant body temperature and characterized by the presence of feathers.  \n",
       "1                A normal or pathological part of the anatomy or structural organization of an organism.  \n",
       "2    An abnormal structure, or one that is abnormal in size or location, present at birth or evolving...  \n",
       "3    An anatomical structure in a fully formed organism; in mammals, for example, a structure in the ...  \n",
       "4    A clinically significant dysfunction whose major manifestation is behavioral or psychological. T...  \n",
       "..                                                                                                   ...  \n",
       "122  A chemical viewed from the perspective of its functional characteristics or pharmacological acti...  \n",
       "123  A biologically active substance whose activities affect or play a role in the functioning of the...  \n",
       "124     A concept which is of interest because it pertains to the carrying out of a process or activity.  \n",
       "125  A conceptual entity resulting from human endeavor. Concepts assigned to this type generally refe...  \n",
       "126  A specific structure or site on the cell surface or within its cytoplasm that recognizes and bin...  \n",
       "\n",
       "[127 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "query = \"SELECT DISTINCT UI, STY_RL, DEF FROM {0}.SRDEF WHERE RT = 'STY'\".format(UMLSversion)\n",
    "df = pd.read_sql_query(query, engine)\n",
    "df.columns =['id', 'name', 'def']\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    f.write(',{\"labels\":[\"Node_Label\"],\"properties\":{\"id\":\"UMLS:'+row[\"id\"]+'\",\"def\":'+json.dumps(row[\"def\"])+',\"node_label\":'+json.dumps(row[\"name\"])+',\"sab\":\"UMLS\"}}\\n')\n",
    "\n",
    "f.close()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Rel_Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rel_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>adjacent_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>after</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>allele_has_abnormality</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>allele_has_activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>allele_in_chromosomal_location</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>uses_excluded_substance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>uses_possibly_included_substance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>uses_substance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>value_set_is_paired_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>was_a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>485 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            rel_label\n",
       "423                       adjacent_to\n",
       "458                             after\n",
       "405            allele_has_abnormality\n",
       "330               allele_has_activity\n",
       "322    allele_in_chromosomal_location\n",
       "..                                ...\n",
       "426           uses_excluded_substance\n",
       "375  uses_possibly_included_substance\n",
       "62                     uses_substance\n",
       "351          value_set_is_paired_with\n",
       "20                              was_a\n",
       "\n",
       "[485 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This could be improved and other options such as use of Relations Ontology and/or other common relationships could be integrated and properly identified, however from the graph performance standpoint, as UMLS alone, only the rel_label will matter per se - Rel_Labels only need ids if they themselves are used as nodes in a semantic network that one wishes to be in the graph - as one might assemble from RO similar to the approach of Node_Label isa relationships inserted later from UMLS Semantic Network\n",
    "\n",
    "df = pd.DataFrame(concept_rels['rel_label'].unique())\n",
    "df.columns = ['rel_label']\n",
    "df.sort_values(by='rel_label', inplace=True)\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    f.write(',{\"labels\":[\"Rel_Label\"],\"properties\":{\"id\":\"UMLS:'+row[\"rel_label\"]+'\",\"def\":\"'+row[\"rel_label\"]+'\",\"rel_label\":\"'+row[\"rel_label\"]+'\",\"sab\":\"UMLS\"}}\\n')\n",
    "        \n",
    "f.close()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Concept nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>pref_term</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C0000275</td>\n",
       "      <td>2-Chloroadenosine</td>\n",
       "      <td>[\"Concept\", \"Nucleic Acid, Nucleoside, or Nucleotide\", \"Pharmacologic Substance\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0000588</td>\n",
       "      <td>Methylbufotenin</td>\n",
       "      <td>[\"Concept\", \"Biologically Active Substance\", \"Organic Chemical\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0000858</td>\n",
       "      <td>Abstracting and Indexing</td>\n",
       "      <td>[\"Concept\", \"Occupational Activity\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C0000973</td>\n",
       "      <td>acetanilide</td>\n",
       "      <td>[\"Concept\", \"Organic Chemical\", \"Pharmacologic Substance\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0001621</td>\n",
       "      <td>Adrenal Gland Diseases</td>\n",
       "      <td>[\"Concept\", \"Disease or Syndrome\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439052</th>\n",
       "      <td>C6020935</td>\n",
       "      <td>Cholesterol.in HDL:{Measurement}:-:Ser/Plas:-</td>\n",
       "      <td>[\"Concept\", \"Clinical Attribute\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439053</th>\n",
       "      <td>C6021227</td>\n",
       "      <td>Coxsackievirus B2 Ab:{Measurement}:-:CSF:-</td>\n",
       "      <td>[\"Concept\", \"Clinical Attribute\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439054</th>\n",
       "      <td>C6021815</td>\n",
       "      <td>Rosmarinus officinalis Ab.IgG&amp;#x7C;ACnc&amp;#x7C;Pt&amp;#x7C;ANYBldSerPl</td>\n",
       "      <td>[\"Concept\", \"Amino Acid, Peptide, or Protein\", \"Immunologic Factor\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439055</th>\n",
       "      <td>C6022125</td>\n",
       "      <td>Radish (Raphanus sativus) IgG &amp;#x7C; Serum or Plasma &amp;#x7C; Allergy</td>\n",
       "      <td>[\"Concept\", \"Amino Acid, Peptide, or Protein\", \"Immunologic Factor\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3439056</th>\n",
       "      <td>C6022232</td>\n",
       "      <td>LOINC, Ukrainian, Ukraine Edition, 281</td>\n",
       "      <td>[\"Concept\", \"Intellectual Product\"]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3439057 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  \\\n",
       "0        C0000275   \n",
       "1        C0000588   \n",
       "2        C0000858   \n",
       "3        C0000973   \n",
       "4        C0001621   \n",
       "...           ...   \n",
       "3439052  C6020935   \n",
       "3439053  C6021227   \n",
       "3439054  C6021815   \n",
       "3439055  C6022125   \n",
       "3439056  C6022232   \n",
       "\n",
       "                                                                   pref_term  \\\n",
       "0                                                          2-Chloroadenosine   \n",
       "1                                                            Methylbufotenin   \n",
       "2                                                   Abstracting and Indexing   \n",
       "3                                                                acetanilide   \n",
       "4                                                     Adrenal Gland Diseases   \n",
       "...                                                                      ...   \n",
       "3439052                        Cholesterol.in HDL:{Measurement}:-:Ser/Plas:-   \n",
       "3439053                           Coxsackievirus B2 Ab:{Measurement}:-:CSF:-   \n",
       "3439054     Rosmarinus officinalis Ab.IgG&#x7C;ACnc&#x7C;Pt&#x7C;ANYBldSerPl   \n",
       "3439055  Radish (Raphanus sativus) IgG &#x7C; Serum or Plasma &#x7C; Allergy   \n",
       "3439056                               LOINC, Ukrainian, Ukraine Edition, 281   \n",
       "\n",
       "                                                                                    labels  \n",
       "0        [\"Concept\", \"Nucleic Acid, Nucleoside, or Nucleotide\", \"Pharmacologic Substance\"]  \n",
       "1                         [\"Concept\", \"Biologically Active Substance\", \"Organic Chemical\"]  \n",
       "2                                                     [\"Concept\", \"Occupational Activity\"]  \n",
       "3                               [\"Concept\", \"Organic Chemical\", \"Pharmacologic Substance\"]  \n",
       "4                                                       [\"Concept\", \"Disease or Syndrome\"]  \n",
       "...                                                                                    ...  \n",
       "3439052                                                  [\"Concept\", \"Clinical Attribute\"]  \n",
       "3439053                                                  [\"Concept\", \"Clinical Attribute\"]  \n",
       "3439054               [\"Concept\", \"Amino Acid, Peptide, or Protein\", \"Immunologic Factor\"]  \n",
       "3439055               [\"Concept\", \"Amino Acid, Peptide, or Protein\", \"Immunologic Factor\"]  \n",
       "3439056                                                [\"Concept\", \"Intellectual Product\"]  \n",
       "\n",
       "[3439057 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here we select the prefered term for each concept using UMLS recommended approach using ISPREF,STT,TS from MRCONSO\n",
    "# And we use LISTAGG in Oracle in pattern to generate JSON-like list of labels from MRSTY\n",
    "# By joining this to CUI-pref_terms we can assemble the Concept node information - write takes perhaps 10-15 min around 500 MB\n",
    "query = \"WITH ConceptLabels AS (SELECT CUI, '[\\\"Concept\\\", ' || LISTAGG('\\\"' || STY || '\\\"', ', ') WITHIN GROUP (ORDER BY STY) || ']' AS labels FROM {0}.MRSTY GROUP BY CUI), ConceptTerms AS (SELECT DISTINCT CUI, STR FROM {0}.MRCONSO WHERE ISPREF = 'Y' AND STT = 'PF' AND TS = 'P' AND LAT = 'ENG') SELECT ConceptTerms.CUI, ConceptTerms.STR, ConceptLabels.labels FROM ConceptTerms LEFT JOIN ConceptLabels ON ConceptTerms.CUI = ConceptLabels.CUI\".format(UMLSversion)\n",
    "df = pd.read_sql_query(query, engine)\n",
    "df.columns =['id','pref_term','labels']\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    f.write(',{\"labels\":'+row[\"labels\"]+',\"properties\":{\"id\":\"UMLS:'+row[\"id\"]+'\",\"sab\":\"UMLS\",\"pref_term\":'+json.dumps(row[\"pref_term\"])+'}}\\n')\n",
    "        \n",
    "f.close()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get CODErels from MRCONSO and MRDEF done here due to being used three times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end_id</th>\n",
       "      <th>codeid</th>\n",
       "      <th>sab</th>\n",
       "      <th>tty</th>\n",
       "      <th>start_id</th>\n",
       "      <th>def</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2,2'-Bipyridine</td>\n",
       "      <td>MSH:D015082</td>\n",
       "      <td>MSH</td>\n",
       "      <td>N1</td>\n",
       "      <td>C0000194</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Droxidopa</td>\n",
       "      <td>GS:4782</td>\n",
       "      <td>GS</td>\n",
       "      <td>IN</td>\n",
       "      <td>C0000378</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7,12 Dimethylbenzanthracene</td>\n",
       "      <td>MSH:D015127</td>\n",
       "      <td>MSH</td>\n",
       "      <td>PM</td>\n",
       "      <td>C0000677</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Abbreviated injury scale (assessment scale)</td>\n",
       "      <td>SNOMEDCT_US:273254002</td>\n",
       "      <td>SNOMEDCT_US</td>\n",
       "      <td>FN</td>\n",
       "      <td>C0000722</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Griping abdominal</td>\n",
       "      <td>MDR:10018727</td>\n",
       "      <td>MDR</td>\n",
       "      <td>LLT</td>\n",
       "      <td>C0000729</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8831514</th>\n",
       "      <td>Neurochondrin Ab:{Measurement}:-:Ser/Plas:-</td>\n",
       "      <td>LNC:108842-6</td>\n",
       "      <td>LNC</td>\n",
       "      <td>LN</td>\n",
       "      <td>C6021186</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8831515</th>\n",
       "      <td>Norbuprenorphine Msmt SerPl</td>\n",
       "      <td>LNC:110228-4</td>\n",
       "      <td>LNC</td>\n",
       "      <td>OSN</td>\n",
       "      <td>C6021246</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8831516</th>\n",
       "      <td>Entity Multimedia file information panel</td>\n",
       "      <td>LNC:107130-7</td>\n",
       "      <td>LNC</td>\n",
       "      <td>LC</td>\n",
       "      <td>C6021512</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8831517</th>\n",
       "      <td>Guidance for balloon obstruction:Finding:To identify measures at a point in time:Abdomen+Pelvis&gt;...</td>\n",
       "      <td>LNC:107406-1</td>\n",
       "      <td>LNC</td>\n",
       "      <td>MTH_LN</td>\n",
       "      <td>C6021542</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8831518</th>\n",
       "      <td>Medication Management Note:Finding:To identify measures at a point in time:{Setting}:Document:Pu...</td>\n",
       "      <td>LNC:107921-9</td>\n",
       "      <td>LNC</td>\n",
       "      <td>MTH_LN</td>\n",
       "      <td>C6021580</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8831519 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                      end_id  \\\n",
       "0                                                                                            2,2'-Bipyridine   \n",
       "1                                                                                                  Droxidopa   \n",
       "2                                                                                7,12 Dimethylbenzanthracene   \n",
       "3                                                                Abbreviated injury scale (assessment scale)   \n",
       "4                                                                                          Griping abdominal   \n",
       "...                                                                                                      ...   \n",
       "8831514                                                          Neurochondrin Ab:{Measurement}:-:Ser/Plas:-   \n",
       "8831515                                                                          Norbuprenorphine Msmt SerPl   \n",
       "8831516                                                             Entity Multimedia file information panel   \n",
       "8831517  Guidance for balloon obstruction:Finding:To identify measures at a point in time:Abdomen+Pelvis>...   \n",
       "8831518  Medication Management Note:Finding:To identify measures at a point in time:{Setting}:Document:Pu...   \n",
       "\n",
       "                        codeid          sab     tty  start_id   def  \n",
       "0                  MSH:D015082          MSH      N1  C0000194  None  \n",
       "1                      GS:4782           GS      IN  C0000378  None  \n",
       "2                  MSH:D015127          MSH      PM  C0000677  None  \n",
       "3        SNOMEDCT_US:273254002  SNOMEDCT_US      FN  C0000722  None  \n",
       "4                 MDR:10018727          MDR     LLT  C0000729  None  \n",
       "...                        ...          ...     ...       ...   ...  \n",
       "8831514           LNC:108842-6          LNC      LN  C6021186  None  \n",
       "8831515           LNC:110228-4          LNC     OSN  C6021246  None  \n",
       "8831516           LNC:107130-7          LNC      LC  C6021512  None  \n",
       "8831517           LNC:107406-1          LNC  MTH_LN  C6021542  None  \n",
       "8831518           LNC:107921-9          LNC  MTH_LN  C6021580  None  \n",
       "\n",
       "[8831519 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This query is used both to subset to unique Terms as nodes and then later for the actual CODE rels\n",
    "query = \"SELECT DISTINCT {0}.MRCONSO.STR, ({0}.MRCONSO.SAB||':'||{0}.MRCONSO.CODE), {0}.MRCONSO.SAB, {0}.MRCONSO.TTY, {0}.MRCONSO.CUI, {0}.MRDEF.DEF FROM {0}.MRCONSO LEFT JOIN {0}.MRDEF ON {0}.MRCONSO.AUI = {0}.MRDEF.AUI WHERE {0}.MRCONSO.LAT = 'ENG' AND {0}.MRCONSO.SUPPRESS <> 'O'\".format(UMLSversion)\n",
    "CODEs = pd.read_sql_query(query, engine)\n",
    "CODEs.columns = ['end_id','codeid','sab','tty','start_id','def']\n",
    "\n",
    "# Code conversiion for codeids with embedded sab such as GO:GO:12345\n",
    "CODEs['codeid'] = [re.split(':', value)[0] + \":\" + re.split(':', value)[-1] for value in CODEs['codeid']]\n",
    "\n",
    "# These character replacements are due to some UMLS codes having these (non-CURIE-like) characters and HL7 having individual-character codes\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace(',', '_', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace('/', '_', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace(' ', '_', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace('<', '__', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace('>', '_', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace('+', '-', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace('*', '-', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace('&', '.', regex=False)\n",
    "CODEs['codeid'] = CODEs['codeid'].str.replace('#', '.', regex=False)\n",
    "\n",
    "# Adjust all Terms that match CURIE-like test - add colon at the end\n",
    "CODEs['end_id'] = [value+\":\" if re.fullmatch(r'^[a-zA-Z0-9\\\\._-]+:[a-zA-Z0-9\\\\._-]+$', value) else value for value in CODEs['end_id']]\n",
    "CODEs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Term nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2,2'-Bipyridine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Droxidopa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7,12 Dimethylbenzanthracene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Abbreviated injury scale (assessment scale)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Griping abdominal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7585471</th>\n",
       "      <td>Hepatitis C virus RNA:{Measurement}:-:Ser/Plas:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7585472</th>\n",
       "      <td>Neurochondrin Ab:{Measurement}:-:Ser/Plas:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7585473</th>\n",
       "      <td>Norbuprenorphine Msmt SerPl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7585474</th>\n",
       "      <td>Guidance for balloon obstruction:Finding:To identify measures at a point in time:Abdomen+Pelvis&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7585475</th>\n",
       "      <td>Medication Management Note:Finding:To identify measures at a point in time:{Setting}:Document:Pu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7585476 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                           0\n",
       "0                                                                                            2,2'-Bipyridine\n",
       "1                                                                                                  Droxidopa\n",
       "2                                                                                7,12 Dimethylbenzanthracene\n",
       "3                                                                Abbreviated injury scale (assessment scale)\n",
       "4                                                                                          Griping abdominal\n",
       "...                                                                                                      ...\n",
       "7585471                                                     Hepatitis C virus RNA:{Measurement}:-:Ser/Plas:-\n",
       "7585472                                                          Neurochondrin Ab:{Measurement}:-:Ser/Plas:-\n",
       "7585473                                                                          Norbuprenorphine Msmt SerPl\n",
       "7585474  Guidance for balloon obstruction:Finding:To identify measures at a point in time:Abdomen+Pelvis>...\n",
       "7585475  Medication Management Note:Finding:To identify measures at a point in time:{Setting}:Document:Pu...\n",
       "\n",
       "[7585476 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here we select the unique end_id among all CODEs and make them Term nodes - this may take a while perhaps 10-15 minute - approx 600 MB\n",
    "\n",
    "df = pd.DataFrame(CODEs['end_id'].unique())\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    f.write(',{\"labels\":[\"Term\"],\"properties\":{\"id\":'+json.dumps(row[0])+'}}\\n')\n",
    "              \n",
    "f.close()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Rel list opening"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we add the Rel list opening to the file\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "f.write('],\"rels\":[\\n')\n",
    "              \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Semantic relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>end_id</th>\n",
       "      <th>start_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T204</td>\n",
       "      <td>T002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T001</td>\n",
       "      <td>T004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T071</td>\n",
       "      <td>T004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T072</td>\n",
       "      <td>T010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T204</td>\n",
       "      <td>T010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>T051</td>\n",
       "      <td>T191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>T071</td>\n",
       "      <td>T194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>T121</td>\n",
       "      <td>T195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>T032</td>\n",
       "      <td>T201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>T071</td>\n",
       "      <td>T204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>464 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    end_id start_id\n",
       "0     T204     T002\n",
       "1     T001     T004\n",
       "2     T071     T004\n",
       "3     T072     T010\n",
       "4     T204     T010\n",
       "..     ...      ...\n",
       "459   T051     T191\n",
       "460   T071     T194\n",
       "461   T121     T195\n",
       "462   T032     T201\n",
       "463   T071     T204\n",
       "\n",
       "[464 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here we add the Semantic relationships, adding first one without a leading comma\n",
    "\n",
    "query = \"WITH Semantics as (SELECT DISTINCT UI from {0}.SRDEF WHERE RT = 'STY') SELECT DISTINCT UI3, UI1 FROM {0}.SRSTRE1 INNER JOIN Semantics ON {0}.SRSTRE1.UI1 = Semantics.UI WHERE UI2 = 'T186'\".format(UMLSversion)\n",
    "df = pd.read_sql_query(query, engine)\n",
    "df.columns =['end_id', 'start_id']\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    if index == 0:\n",
    "        f.write('{\"label\":\"isa\",\"end\":{\"properties\":{\"id\":\"UMLS:'+row[\"end_id\"]+'\"}},\"properties\":{\"sab\":\"UMLS\"},\"start\":{\"properties\":{\"id\":\"UMLS:'+row[\"start_id\"]+'\"}}}\\n')\n",
    "    else:\n",
    "        f.write(',{\"label\":\"isa\",\"end\":{\"properties\":{\"id\":\"UMLS:'+row[\"end_id\"]+'\"}},\"properties\":{\"sab\":\"UMLS\"},\"start\":{\"properties\":{\"id\":\"UMLS:'+row[\"start_id\"]+'\"}}}\\n')\n",
    "              \n",
    "f.close()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Concept-Concept relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we add the Concept-Concept relationships queried earlier - this adds about 1.7 Gigabyte to the file\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "for index, row in concept_rels.iterrows():\n",
    "    f.write(',{\"label\":\"'+row[\"rel_label\"]+'\",\"end\":{\"properties\":{\"id\":\"UMLS:'+row[\"end_id\"]+'\"}},\"properties\":{\"sab\":\"'+row[\"sab\"]+'\"},\"start\":{\"properties\":{\"id\":\"UMLS:'+row[\"start_id\"]+'\"}}}\\n')\n",
    "              \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add CODE relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we add the CODE relationships queried earlier - this adds about 1.7 Gigabytes to the file\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "for index, row in CODEs.iterrows():\n",
    "    f.write(',{\"label\":\"CODE\",\"end\":{\"properties\":{\"id\":'+json.dumps(row[\"end_id\"])+'}},\"properties\":{\"sab\":\"'+row[\"sab\"]+'\",\"def\":'+json.dumps(row[\"def\"])+',\"tty\":\"'+row[\"tty\"]+'\",\"codeid\":\"'+row[\"codeid\"]+'\"},\"start\":{\"properties\":{\"id\":\"UMLS:'+row[\"start_id\"]+'\"}}}\\n')\n",
    "              \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NDC relation to RXNORM - make NDC an SAB and do ingest of CODEs parallel to RXNORM ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>codeid</th>\n",
       "      <th>end_id</th>\n",
       "      <th>sab</th>\n",
       "      <th>tty</th>\n",
       "      <th>start_id</th>\n",
       "      <th>def</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NDC:65219047905</td>\n",
       "      <td>calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SCD</td>\n",
       "      <td>C0140600</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NDC:50989089832</td>\n",
       "      <td>calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SCD</td>\n",
       "      <td>C0140600</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NDC:72483020201</td>\n",
       "      <td>calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SCD</td>\n",
       "      <td>C0140600</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NDC:86156080360</td>\n",
       "      <td>calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SCD</td>\n",
       "      <td>C0140600</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NDC:65219047530</td>\n",
       "      <td>calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SCD</td>\n",
       "      <td>C0140600</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824719</th>\n",
       "      <td>NDC:50222028091</td>\n",
       "      <td>delgocitinib 20 MG/ML Topical Cream [Anzupgo]</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SBD</td>\n",
       "      <td>C6016980</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824724</th>\n",
       "      <td>NDC:68727025001</td>\n",
       "      <td>dordaviprone 125 MG Oral Capsule [Modeyso]</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SBD</td>\n",
       "      <td>C6016981</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824726</th>\n",
       "      <td>NDC:15370050116</td>\n",
       "      <td>celecoxib 10 MG/ML Oral Suspension [Vyscoxa]</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SBD</td>\n",
       "      <td>C6017233</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824730</th>\n",
       "      <td>NDC:73362001602</td>\n",
       "      <td>lonapegsomatropin-tcgd 2.5 MG Cartridge [Skytrofa]</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SBD</td>\n",
       "      <td>C6017308</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824732</th>\n",
       "      <td>NDC:71558000230</td>\n",
       "      <td>brensocatib 25 MG Oral Tablet [Brinsupri]</td>\n",
       "      <td>RXNORM</td>\n",
       "      <td>SBD</td>\n",
       "      <td>C6017337</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250589 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 codeid  \\\n",
       "4       NDC:65219047905   \n",
       "9       NDC:50989089832   \n",
       "14      NDC:72483020201   \n",
       "19      NDC:86156080360   \n",
       "24      NDC:65219047530   \n",
       "...                 ...   \n",
       "824719  NDC:50222028091   \n",
       "824724  NDC:68727025001   \n",
       "824726  NDC:15370050116   \n",
       "824730  NDC:73362001602   \n",
       "824732  NDC:71558000230   \n",
       "\n",
       "                                                                                                     end_id  \\\n",
       "4       calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...   \n",
       "9       calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...   \n",
       "14      calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...   \n",
       "19      calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...   \n",
       "24      calcium chloride 0.2 MG/ML / potassium chloride 0.3 MG/ML / sodium chloride 6 MG/ML / sodium lac...   \n",
       "...                                                                                                     ...   \n",
       "824719                                                        delgocitinib 20 MG/ML Topical Cream [Anzupgo]   \n",
       "824724                                                           dordaviprone 125 MG Oral Capsule [Modeyso]   \n",
       "824726                                                         celecoxib 10 MG/ML Oral Suspension [Vyscoxa]   \n",
       "824730                                                   lonapegsomatropin-tcgd 2.5 MG Cartridge [Skytrofa]   \n",
       "824732                                                            brensocatib 25 MG Oral Tablet [Brinsupri]   \n",
       "\n",
       "           sab  tty  start_id   def  \n",
       "4       RXNORM  SCD  C0140600  None  \n",
       "9       RXNORM  SCD  C0140600  None  \n",
       "14      RXNORM  SCD  C0140600  None  \n",
       "19      RXNORM  SCD  C0140600  None  \n",
       "24      RXNORM  SCD  C0140600  None  \n",
       "...        ...  ...       ...   ...  \n",
       "824719  RXNORM  SBD  C6016980  None  \n",
       "824724  RXNORM  SBD  C6016981  None  \n",
       "824726  RXNORM  SBD  C6017233  None  \n",
       "824730  RXNORM  SBD  C6017308  None  \n",
       "824732  RXNORM  SBD  C6017337  None  \n",
       "\n",
       "[250589 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# may wish to run an assessment to see what RXNORM term types these include (unique list on inner join with CODE rels)\n",
    "# then decide how to handle adding more CODE rels of these (before closing) - inner join and replace ndc for rxnorm...\n",
    "# don't use these TTYs: SY, PSN, TMSY\n",
    "\n",
    "query = \"SELECT DISTINCT (SAB||':'||CODE), ('NDC:'||ATV) FROM {0}.MRSAT WHERE SAB = 'RXNORM' and ATN = 'NDC' and SUPPRESS <> 'O'\".format(UMLSversion)\n",
    "df = pd.read_sql_query(query, engine)\n",
    "df.columns =['codeid','ndcid']\n",
    "df = pd.merge(df, CODEs, how='left')\n",
    "df = df[~df['tty'].isin(['SY','PSN','TMSY'])]\n",
    "df.drop('codeid', axis=1, inplace=True)\n",
    "df.rename(columns={'ndcid': 'codeid'}, inplace=True)\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    f.write(',{\"label\":\"CODE\",\"end\":{\"properties\":{\"id\":'+json.dumps(row[\"end_id\"])+'}},\"properties\":{\"sab\":\"NDC\",\"tty\":\"'+row[\"tty\"]+'\",\"codeid\":\"'+row[\"codeid\"]+'\"},\"start\":{\"properties\":{\"id\":\"UMLS:'+row[\"start_id\"]+'\"}}}\\n')\n",
    "              \n",
    "f.close()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add closing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we add the Rel list closing bracket and final closing curly brace to the file\n",
    "\n",
    "f = open('UMLS-JKG.json', 'a')\n",
    "\n",
    "f.write(']}')\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
